Start adding support for Bonding Ethernet and Wifi

skilau opened the pull request:
NOTE: This is NOT complete yet!

This patch adds support for Bonding the Ethernet and Wifi interfaces as an "active-backup".
(It will default to using the Ethernet interface first, and if disconnected for some reason, will fall back to Wifi)

NOTE: The Ethernet works fine, and when disconnected, Wifi is attempted to be called.
However, when it attempts to flip over to Wifi, it can't, because it does not have the wpa_supplicant info to do it.

I am unfamiliar with how wpa_supplicant works, so it is unclear to me how to feed the /tmp/wpa_supplicant.conf file
to the Wifi interface properly.
I am hoping someone more familiar with it, will be able to run with this.

Below is an example of what we see when the Ethernet cable is connected, disconnected, reconnected:

[  461.155016] bonding: wlan0: link status down for active interface eth0, disabling it in 5000 ms.
[  466.155060] bonding: wlan0: link status definitely down for interface eth0, disabling it
[  466.163594] bonding: wlan0: now running without any active interface !
[  507.040168] asix 1-1:1.0 eth0: link up, 10Mbps, half-duplex, lpa 0x0000
[  507.075075] bonding: wlan0: link status up for interface eth0, enabling it in 0 ms.
[  507.093796] bonding: wlan0: link status definitely up for interface eth0, 10 Mbps half duplex.
[  507.102779] bonding: wlan0: making interface eth0 the new active one.
[  507.115347] bonding: wlan0: first active interface up!

gtxaspec commented:
@skilau try
`wpa_supplicant -D nl80211 -i wlan0 -c /tmp/wpa_supplicant.conf -B -s`

gtxaspec commented:
you should also add logic to rename the interface back to wlan0, and unbind wpa_cli.sh to keep iCamera happy. (if you haven't already)

skilau commented:
@gtxaspec : That works!  Thanks!

I leave the "real" wlan0 as your name of wlanold, so I just had to do:
wpa_supplicant -D nl80211 -i wlanold -c /tmp/wpa_supplicant.conf -B -s

Yanking the Ethernet cable now sends us back to Wifi, and continues to work (after a slight delay of course), as though nothing happened.
When I connect the Ethernet cable back, it continues back on Ethernet, as though nothing happened.

[ 4986.344982] bonding: wlan0: link status up for interface wlanold, enabling it in 5000 ms.
[ 4991.346983] bonding: wlan0: link status definitely up for interface wlanold, 0 Mbps full duplex.
[ 5012.055010] bonding: wlan0: link status down for active interface eth0, disabling it in 5000 ms.
[ 5017.058084] bonding: wlan0: link status definitely down for interface eth0, disabling it
[ 5017.066484] bonding: wlan0: making interface wlanold the new active one.
[ 5017.855234] [atbm_log]:ieee80211_process_addba_request:mac[32:94:ed:fc:1d:63],tid[0]
[ 5155.257609] asix 1-1:1.0 eth0: link up, 100Mbps, full-duplex, lpa 0xC5E1
[ 5155.285067] bonding: wlan0: link status up for interface eth0, enabling it in 5000 ms.
[ 5160.285018] bonding: wlan0: link status definitely up for interface eth0, 10 Mbps half duplex.
[ 5160.293942] bonding: wlan0: making interface eth0 the new active one.

I will clean some things up, do some more testing, and resubmit the patch in the next couple days.
Thanks again!

claudobahn commented:
@skilau looks like whatever you're using to edit the scripts is replacing spaces w/ tabs, you'll probably want to make sure you stick w/ the kind of white space that's already in the scripts.

skilau commented:
@claudobahn: Yeah, I was looking at the script, to decide what the standard was, and it appears to currently be a mix of both tabs and spaces, so I wasn't sure which one to go with.
I went with tabs.  I will switch to spaces instead.  :)

claudobahn commented:
Ah, I only pointed it out because I noticed white space changes to lines that had no other changes and those changes turned out to be spaces replaced w/ tabs.

No idea what @gtxaspec prefers :)

claudobahn commented:
Ah, yes, I see what you mean. Some lines have both, even :) So I suppose it doesn't matter.

gtxaspec commented:
that's something I need to fix, my terminal converts tabs to spaces when I copy... I usually prefer tabs though.

skilau commented:
No problem, I reverted the lines I changed back to whatever they were.
(I typically do real tabs for shell-type scripts, and do spaces for Python and C)

skilau commented:
Finally, this patch, at least for me, works as I had hoped originally now.
It prioritizes Ethernet and uses that if at all possible.  If the cable is yanked or something is wrong, the bonding will flip over to Wifi and continue as though nothing happened.
If Ethernet comes back, it will switch back, without any interruption.

If you pull the cable while watching the image in the Wyze app, you can see the 5 second delay while "bonding" detects the cable being gone, and then the image will catch up as it flips over to Wifi, without disrupting the flow of data/traffic.

Beyond a bit more testing I want to do...

Presumably I should have a top level variable, maybe called something like ENABLE_INTERFACE_BONDING true/false, but only if Ethernet is enabled, as maybe not everyone would want this feature.
So I will be adding that in the next couple days when I get some time.

Thanks again!

skilau commented:
OK, @gtxaspec and @claudobahn, I am reasonably happy with this pull request now.

I was able to test that it works properly when:
bonding=true, primary=eth0, secondary=wlan0
bonding=true, primary=wlan0, secondary=eth0
bonding=true, eth0 = false (We won't do bonding, because we need/want eth0 to do it)
bonding=false, eth0 = true (We won't do bonding, because bonding is off)
bonding=false, eth0 = false (We won't do bonding, because bonding and eth is off)

I defaulted bonding to false, so it won't surprise anyone by mistake.

Let me know if this is an acceptable patch, and/or you need/want me to change anything more!

Thanks again!

gtxaspec commented:
can you rebase off the latest release? wz_user.sh changed a bunch of variables.

skilau commented:
> can you rebase off the latest release? wz_user.sh changed a bunch of variables.

No problem!   Done.  Thanks!

gtxaspec commented:
@skilau wrapping up some stuff here and I will have time to look at your pull soon.  Just a question, does this work for usb-direct too (usb0 interface name)

skilau commented:
Hi @gtxaspec,
No problem at all.  Your project, your time. :)

I have not tried it against the USB-Direct interface, as I don't have a set up here to be able to try it and see.

As far as I can tell, in theory it should work against it, as the "binding" driver only says that anything that has something-like an Ethernet interface should work with it.

gtxaspec commented:
@skilau ok so I did some testing with your branch today, and it works If my ethernet adapter and wlan are on the same subnet, but it doesn't if they are on different subnets.  Do you know of how to fix this? Do we need to run dhcp again? restart the supplicant?

gtxaspec commented:
@skilau ok, figured this out:

`kill -USR2 <udhcpc pid>` will force the dhcp client to send a DHCP RELEASE, and then `kill -USR1 <udhcpc pid>` forces the client to DHCP RENEW.  How can we trigger this command when the eth0 interface is down...

gtxaspec commented:
@skilau, merged!

Would you mind updating the README too? 

thanks

skilau commented:
Awesome, thank you!
I did another small pull request to add the update to the README as requested!
Thanks again!
